### PR TITLE
Add lazy evaluation feature to Tensor @open sesame 06/09 16:16

### PR DESCRIPTION
**Pull request by zhoonit**
This PR propose 2 concepts.

1. Propose `*_i` type of operation to `Tensor`. Which mutates target
tensor instead of memcopying the tensor.(511c74d)
2. Add chaining for lazy & memcopyless operation.

For example Tensor could do the operation in such manner:

```cpp
Tensor t;
t.chain() /* Initial memcpy happens to guarantee immutability */
 .add_i(x)
 .multiply_i(y) /* NYI */
 .divide_i(y) /* NYI */
 .run()
```

operations are delayed until `.run()` is called. This is done by
monadic object named `DefferedTensor`.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

File: nntrainer/include/tensor.h [hunk 1]
```diff
@@ -41,6 +41,7 @@ extern "C" {
 
 namespace nntrainer {
 
+class DeferredTensor;
```
Comment (jijoongmoon) [inline nntrainer/include/tensor.h:44]: I think this looks like cyclic dependency. How about implement chain and run inside tensor and split math operation into separate file something like tensor_op.h?

File: nntrainer/include/tensor.h [hunk 2]
```diff
@@ -41,6 +41,7 @@ extern "C" {
...
 /**
```
Comment (jijoongmoon) [inline nntrainer/include/tensor.h:45]: how about implement tensor directly. Cause if this works well, we are not going to use tensor any way. Isn't it? and also it looks like duplication to have add and add_i.

File: nntrainer/include/tensor.h (hunk 1 above)
Comment (jijoongmoon) [inline nntrainer/include/tensor.h:44]: how about implement tensor directly. Cause if this works well, we are not going to use tensor any way. Isn't it? and also it looks like duplication to have add and add_i

Comment (zhoonit) [reply nntrainer/include/tensor.h:44]: 1. About deprecating existing operator (change `add_i` to `add`):

I think this could be done after we change all traditional operation used in the code base and be sure that the traditional version is not needed in all case since mechanism is slightly different. 

for example:
```cpp
 Tensor t = t.add(3.14);
``` 
won't work if we change now. 

2. Combining this to tensor:

IMHO it is better to leave monadic wrapper to have a separate class handling lazy evaluation if we are going to use the same method name. 

If we migrate this to tensor, we need 1. a status added to tell if it is lazy evaluating and 2. case handling for every operation and 3. same code bits in every operation methods.

more specifically this could be done like this but IMO less practical:
```cpp
Tensor Tensor::some_operation(float const &value) {
 /* code bits that should be added for every operation */
  if(this->chained) {
    call_chain.push_back([value](Tensor &t) mutable -> int {
      this->some_operation(value);
      return ML_ERROR_NONE;
    });
    return *this;
  }
 /* handle real operations */
}
```

As for as I can tell, I think it is better to leave as it is for now.
Please tell me if those can be implemented in better way!

Comment (jijoongmoon) [reply nntrainer/include/tensor.h:44]: ```Tensor t = t.add(3.14); ``` I think we are not going to use this after this deferred tensor is enabled. May be use this ``` Tensor t = t.chain().add(3.14).run()```. And when we do the chain(), it copy the tensor. I think it looks like tensor has tensor copy and do the function which is added by the numeric function ( current functions like add() is not going to use. and if we update add_i to add, then it add function into call_chain. and run() and save the data into copied tensor and may be update the original if there is assign operation. How about it?

Comment (jijoongmoon) [reply nntrainer/include/tensor.h:44]: We had agreement after some valuable discussions. Yes. we are going to use this approach. But change Class name with LazyTensor to make more clear. Thanks.

Comment (zhoonit) [reply nntrainer/include/tensor.h:44]: Thanks for the suggestion :) I'll change the class name and rebase other PR 👍

File: nntrainer/include/lazy_tensor.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+#ifdef __cplusplus
+
+#include <tensor.h>
+#include <vector>
+
+namespace nntrainer {
+
+/**
+ * @class   LazyTensor a wrapper class for lazy calculation of tensor
+ * @brief   calculation is delayed until Tensor DefferedTensor::run() is
```
Comment (dongju-chae) [inline nntrainer/include/lazy_tensor.h:24]: Outdated `DefferedTensor`?

File: nntrainer/include/tensor.h [hunk 4]
```diff
@@ -110,13 +111,28 @@ class Tensor {
...
                  unsigned int w);
 
+  /**
+   * @brief     Scalar multiplication and apply immediatly
+   * @param[in] value Scalar value to be multiplied
+   * @retval    #ML_ERROR_NONE  Successful
+   * @retval    #ML_ERROR_INVALID_PARAMETER Tensor dimension is not right
+   */
+  int multiply_i(float const &value);
+
```
Comment (jijoongmoon) [inline nntrainer/include/tensor.h:121]: I think this is not related with this PR. You would be better to apply it other PR.

File: nntrainer/include/tensor.h [hunk 5]
```diff
@@ -110,13 +111,28 @@ class Tensor {
...
    */
   Tensor multiply(float const &value);
 
+  /**
+   * @brief     Scalar divide and apply immediatly   * @param[in] value Scalar value to be divided
+   * @retval    #ML_ERROR_NONE  Successful
+   * @retval    #ML_ERROR_INVALID_PARAMETER Tensor dimension is not right
+   */
+  int divide_i(float const &value);
+
```
Comment (jijoongmoon) [inline nntrainer/include/tensor.h:135]: this is same with line no.121

File: nntrainer/include/tensor.h [hunk 6]
```diff
@@ -160,13 +176,30 @@ class Tensor {
...
 
+
+  /**
+   * @brief     Multiply Tensor Elementwise and apply immediatly
+   * @param[in] m Tensor to be multiplied
+   * @retval    #ML_ERROR_NONE  Successful
+   * @retval    #ML_ERROR_INVALID_PARAMETER Tensor dimension is not right
+   */
+  int multiply_i(Tensor const &m) const;
+
```
Comment (jijoongmoon) [inline nntrainer/include/tensor.h:187]: this is same with line no.121

File: nntrainer/include/tensor.h [hunk 7]
```diff
@@ -160,13 +176,30 @@ class Tensor {
...
   Tensor multiply(Tensor const &m) const;
 
+ /**
+   * @brief     Divide Tensor Elementwise and apply immediatly
+   * @param[in] m Tensor to be multiplied
+   * @retval    #ML_ERROR_NONE  Successful
+   * @retval    #ML_ERROR_INVALID_PARAMETER Tensor dimension is not right
+   */
+  int divide_i(Tensor const &m) const;
+
```
Comment (jijoongmoon) [inline nntrainer/include/tensor.h:202]: this is same with line no.121

File: nntrainer/include/tensor.h (hunk 4 above)
Comment (zhoonit) [reply nntrainer/include/tensor.h]: Thanks. It is updated :)

File: nntrainer/include/tensor.h (hunk 5 above)
Comment (zhoonit) [reply nntrainer/include/tensor.h]: Thank you! it is deleted

File: nntrainer/include/tensor.h (hunk 6 above)
Comment (zhoonit) [reply nntrainer/include/tensor.h]: Thank you! it is deleted!

File: nntrainer/include/tensor.h (hunk 7 above)
Comment (zhoonit) [reply nntrainer/include/tensor.h]: Thank you! it is deleted

File: test/unittest/unittest_nntrainer_lazy_tensor.cpp [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+TEST(nntrainer_LazyTensor, LazyTensor_01_p) {
+  int batch = 3;
+  int height = 3;
+  int width = 10;
+  int channel = 1;
+
+  nntrainer::Tensor target(batch, channel, height, width);
+  GEN_TEST_INPUT(target, i * (batch * height) + j * (width) + k + 1 + l);
+
+  nntrainer::LazyTensor defed(target);
```
Comment (dongju-chae) [inline test/unittest/unittest_nntrainer_lazy_tensor.cpp:30]: what's meaning of `defed`?

Comment (zhoonit) [reply test/unittest/unittest_nntrainer_lazy_tensor.cpp]: It meant `deferred`. I think the name is unclear, I changed it to `delayed`. Thanks
